image_transport/src/image_transport.cpp

```diff
@@ -44,78 +44,193 @@ namespace image_transport
 
 struct Impl
 {
-  PubLoaderPtr pub_loader_;
-  SubLoaderPtr sub_loader_;
+  PubLoaderPtr<rclcpp::Node> pub_loader_;
+  SubLoaderPtr<rclcpp::Node> sub_loader_;
 
   Impl()
-  : pub_loader_(std::make_shared<PubLoader>("image_transport", "image_transport::PublisherPlugin")),
-    sub_loader_(std::make_shared<SubLoader>("image_transport", "image_transport::SubscriberPlugin"))
+  : pub_loader_(std::make_shared<PubLoader<rclcpp::Node>>(
+        "image_transport",
+        "image_transport::PublisherPlugin<rclcpp::Node>")),
+    sub_loader_(std::make_shared<SubLoader<rclcpp::Node>>(
+        "image_transport",
+        "image_transport::SubscriberPlugin<rclcpp::Node>"))
+  {
+  }
+};
+
+struct ImplLifecycle
+{
+  PubLoaderPtr<rclcpp_lifecycle::LifecycleNode> pub_loader_;
+  SubLoaderPtr<rclcpp_lifecycle::LifecycleNode> sub_loader_;
+
+  ImplLifecycle()
+  : pub_loader_(std::make_shared<PubLoader<rclcpp_lifecycle::LifecycleNode>>(
+        "image_transport",
+        "image_transport::PublisherPlugin<rclcpp_lifecycle::LifecycleNode>")),
+    sub_loader_(std::make_shared<SubLoader<rclcpp_lifecycle::LifecycleNode>>(
+        "image_transport",
+        "image_transport::SubscriberPlugin<rclcpp_lifecycle::LifecycleNode>"))
   {
   }
 };
 
 static Impl * kImpl = new Impl();
+static ImplLifecycle * kImpl_lifecycle = new ImplLifecycle();
 
-Publisher create_publisher(
-  rclcpp::Node * node,
+template<class NodeType>
+Publisher<NodeType> create_publisher(
+  NodeType * node,
   const std::string & base_topic,
   rmw_qos_profile_t custom_qos,
   rclcpp::PublisherOptions options)
 {
-  return Publisher(node, base_topic, kImpl->pub_loader_, custom_qos, options);
+  if constexpr (std::is_same_v<NodeType, rclcpp::Node>) {
+    return Publisher(node, base_topic, kImpl->pub_loader_, custom_qos, options);
+  }
+  if constexpr (std::is_same_v<NodeType, rclcpp_lifecycle::LifecycleNode>) {
+    return Publisher(node, base_topic, kImpl_lifecycle->pub_loader_, custom_qos, options);
+  }
```

Comment:
I wonder if this would be clearer by instead just having two fully specialized functions and avoiding the std::is_same_v

Reply:
The initial implementation of the PR was two specialized functions. I changed it based on https://github.com/ros-perception/image_common/pull/304#pullrequestreview-1991121656 , I'm happy to change it back as well.